Reduce allocation from data taken in local performance trace

sharwell opened the pull request:
* 00b15ac saves over 100MB allocations in 100 seconds
* 1745e06 and 4d8ba5617e8f1ab9cb72d517a37ecf8ae0d776f5 save over 3GB allocations

I can separate refactoring 7e6b174fcd281e5f49f0c8619e952e4844d8d38d if desired.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs:
```diff
@@ -470,45 +455,49 @@ private static int BinarySearch(string concatenatedNames, ImmutableArray<Node> n
+                var members = rootContainer.GetMembers(node.Name);
+                if (!members.IsEmpty)
+                {
+                    results ??= ArrayBuilder<ISymbol>.GetInstance();
```
this sems really odd.  why not just have the array builder created up front and passed in.  that's much more idiomatic with how we do this elsewhere.

sharwell replied:
Mostly because it's slow, particularly for taking an object from the pool and then returning it without using it. More than 750MB of the allocations in this were the pooled objects, while only 50MB of arrays were produced by `ToImmutableAndFree`. This means one of two things is likely happening:

1. The object pool is hitting a high rate of lost objects due to parallel code returning them (this on an 8-core machine and not the 64-core machine I normally test on)
2. The objects are getting held for longer than necessary, leading to more than 128 instances in use at once

sharwell replied:
➡️ Documented the intentional lazy initialization

sharwell replied:
➡️ Transitioned to `TemporaryArray<T>`

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs:
```diff
@@ -580,24 +569,35 @@ internal void AssertEquivalentTo(SymbolTreeInfo other)
-
-            foreach (var derivedTypeIndex in derivedTypeIndices)
+            ArrayBuilder<INamedTypeSymbol>? builder = null;
+            ArrayBuilder<ISymbol>? tempBuilder = null;
```
it feels liek thsi would be much simpler and clearer if we just made two builders here unilaterally.  one to actually store the results, and hte other as a temp scratch buffer that could be passed around and used at sublevels, and cleared when done with.

sharwell replied:
This assumes at least one object will be stored in these builders for the majority of calls. According to performance traces this is not the case, resulting in significant overhead for attempting to obtain and return the builders.

sharwell replied:
➡️ Documented the intentional lazy initialization

sharwell replied:
➡️ Transitioned to `TemporaryArray<T>` which does not need this documentation

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs:
```diff
@@ -181,26 +181,46 @@ public void WriteTo(ObjectWriter writer)
+                    }
+                }
+
+                yield return sortedNodes.AsMemory()[startIndex..sortedNodes.Length];
```
can you doc this method

sharwell replied:
Done

CyrusNajmabadi commented:
One thing i didn't see here was rev'ing any persistence format version so that we won't try to read old data out using your new format.  Or did i misread, and the format is the same, you just changed how we read and write it?

sharwell commented:
> One thing i didn't see here was rev'ing any persistence format version

We should be using AssemblyInformationalVersion for this, but since we aren't I have now increased the version.

sharwell commented:
Marked as draft since I will rebase the `fixup!` commits once the reviews complete.

sharwell commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs:
```diff
@@ -470,45 +456,33 @@ private static int BinarySearch(string concatenatedNames, ImmutableArray<Node> n
+                Bind(node.ParentIndex, rootContainer, ref Unsafe.AsRef(in containerSymbols), cancellationToken);
+#pragma warning disable RS0042 // Do not copy value
                 foreach (var containerSymbol in containerSymbols)
+#pragma warning restore RS0042 // Do not copy value
```
📝 RS0042 doesn't know how to handle `foreach` usage right now, but I'll update it and the suppressions related to this usage will go away.

sharwell replied:
➡️ RS0042 has been updated to handle this

jaredpar commented on src/Compilers/Core/Portable/CaseInsensitiveComparison.cs:
```diff
@@ -117,6 +117,24 @@ public override int Compare(string? str1, string? str2)
                 return str1.Length - str2.Length;
             }
 
+#if NETSTANDARD2_0 || NETCOREAPP
```
Why are we not adding these APIs to desktop? Is one of the unicode APIs that we are using not available there?

sharwell replied:
> Why are we not adding these APIs to desktop?

Desktop uses the **netstandard2.0** target, except for net20 targets that don't support these types.

jaredpar replied:
In that case let's use `#if !NET20` to make it clearer what's happpening here.

sharwell replied:
There are also some netstandard 1.x targets that were impacted, but I can switch to a targeted exclusion.

sharwell replied:
➡️ Switched to

```
#if !NET20 && !NETSTANDARD1_3
```

jaredpar commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.Node.cs:
```diff
@@ -50,9 +49,9 @@ private string GetDebuggerDisplay()
         private struct Node
         {
```
`readonly struct` as long as we're here

sharwell replied:
I've been leaving this for [EPS01](https://github.com/SergeyTeplyakov/ErrorProne.NET/blob/775522226c2d17018e4379c2280d18ddfca06888/src/ErrorProne.NET.StructAnalyzers/MakeStructReadOnlyAnalyzer.cs#L8-L12), either by referencing that analyzer package or porting the rule to dotnet/roslyn-analyzers.

sharwell commented:
@CyrusNajmabadi @jaredpar this is now cleaned up and rebased. The only content changes in the rebase were the final two commits, which can be reviewed per-commit.

@dotnet/roslyn-compiler I will need a second review here in addition to the one from Jared.

AlekseyTs commented on src/Compilers/Core/Portable/PublicAPI.Shipped.txt:
```diff
@@ -356,9 +357,11 @@ Microsoft.CodeAnalysis.CandidateReason.WrongArity = 5 -> Microsoft.CodeAnalysis.
 Microsoft.CodeAnalysis.CaseInsensitiveComparison
 Microsoft.CodeAnalysis.ChildSyntaxList
 Microsoft.CodeAnalysis.ChildSyntaxList.Any() -> bool
+Microsoft.CodeAnalysis.ChildSyntaxList.ChildSyntaxList() -> void
```
I am not sure why this file is getting modified.

sharwell replied:
The new version of the analyzers adds implicit struct constructors (that already existed) to the public API files.

AlekseyTs replied:
> ➡️ The new version of the analyzers adds implicit struct constructors (that already existed) to the public API files. The changes for the analyzer update are in one commit, and a second commit that updates the file to have the lines in the order the analyzer wants (for future maintenance).

I think changes like that should be moved into a separate PR

---
In reply to: [512858730](https://github.com/dotnet/roslyn/pull/48896#discussion_r512858730) [](ancestors = 512858730)

sharwell replied:
I needed a new version of the analyzer to address a false positive in one of the earlier commits. Do you want me to split this pull request? I'm happy to exclude the last two commits with the understanding that a couple of warnings will be disabled in the change until a future PR updates the analyzers.

AlekseyTs replied:
I do not expect a PR that has a goal "Reduce allocation from data taken in local performance trace" to touch any PublicAPI.Shipped files.

---
In reply to: [512863263](https://github.com/dotnet/roslyn/pull/48896#discussion_r512863263) [](ancestors = 512863263)

sharwell replied:
➡️ Will separate to a new PR.

AlekseyTs commented:
> Microsoft.CodeAnalysis.CSharp.AwaitExpressionInfo.AwaitExpressionInfo() -> void

I am not sure why this file is getting modified.

---
Refers to: src/Compilers/CSharp/Portable/PublicAPI.Shipped.txt:63 in a9609cc. [](commit_id = a9609cc5a15f720242563a0271a45dc517830bb7, deletion_comment = False)

AlekseyTs commented:
> Microsoft.CodeAnalysis.VisualBasic.Conversion.New() -> Void

I am not sure why this file is getting modified.

---
Refers to: src/Compilers/VisualBasic/Portable/PublicAPI.Shipped.txt:33 in a9609cc. [](commit_id = a9609cc5a15f720242563a0271a45dc517830bb7, deletion_comment = False)

AlekseyTs commented on src/Dependencies/PooledObjects/ArrayBuilder.cs:
```diff
@@ -281,20 +304,7 @@ public ImmutableArray<U> ToDowncastedImmutable<U>()
         /// </summary>
         public ImmutableArray<T> ToImmutableAndFree()
```
>ToImmutableAndFree [](start = 33, length = 18)

Is this function more expensive now, i.e. does more work than it used to?

sharwell replied:
It appears yes. For builders with a capacity larger than 128 elements, there is a new call to `Clear` before it is discarded. i.e. the call to `Clear` was (unintentionally) hoisted outside this guard:

https://github.com/dotnet/roslyn/blob/a9609cc5a15f720242563a0271a45dc517830bb7/src/Dependencies/PooledObjects/ArrayBuilder.cs#L337

sharwell replied:
➡️ Restored original implementation of `ToImmutableAndFree` with comment to prevent future regression.

AlekseyTs commented:
Done with review pass for changes under Compilers (iteration 8)

AlekseyTs reviewed (approved):
Changes under Compilers LGTM (iteration 6)

sharwell commented:
@jaredpar @CyrusNajmabadi ping

CyrusNajmabadi commented on src/Dependencies/PooledObjects/ArrayBuilder.cs:
```diff
@@ -71,6 +71,29 @@ public ImmutableArray<T> ToImmutable()
+            else
+            {
+                result = ToImmutable();
+                Clear();
```
this path here honestly feels like it masks a bug.  normal ImmutableArray.Builder throws in this case, indicating that the client messed something up.  I think we should do the same.  In other words, MoveToImmutable is suppsoed to be for the case where you know the count and know you've filled all of that, and thus can safely and efficiently use the underlying storage.  This isn't hte case here, so it feels like it's putting different perf expectations onto this call, even though it's called the same thing as the equiv builder operation.

sharwell replied:
This was intentional. The current method is a "move and clear" using the most efficient operation currently available. Roslyn does not use the unintuitive semantics of `ImmutableArray<T>.Builder.MoveToImmutable` and it didn't seem advantageous to introduce complexity with those semantics.

CyrusNajmabadi replied:
Please rename then (maybe ToImmutableAndClear()).  FWIW, i find the semantics of MoveToImmutable very very very intuitive.  It's exactly as it says.  It takes the underlying array and creates the ImmutableArray out of it, doing it as a simple O(1) move instead of O(n) copy.  It's exception behavior is also intuitive and appropriate IMO.  The only safe way to do this is if the count/capacity are paired.  When not, youi've done something wrong and it throws.

So, please either make the semantics the same, or rename.  I'm fine with either. I'm not ok with teh exact same name having different semantics.  Tnx :)

sharwell replied:
I renamed this to `ToImmutableAndClear`. My original objection https://github.com/dotnet/corefx/pull/180#issuecomment-69349402 to the semantics of `MoveToImmutable` has not changed in the time since.

CyrusNajmabadi replied:
Thanks very much :) #resolved

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs:
```diff
@@ -222,18 +214,18 @@ public SymbolTreeInfo WithChecksum(Checksum checksum)
                 assemblySymbol ??= await lazyAssembly.GetValueAsync(cancellationToken).ConfigureAwait(false);
 
-                Bind(node, assemblySymbol.GlobalNamespace, results, cancellationToken);
+                Bind(node, assemblySymbol.GlobalNamespace, ref Unsafe.AsRef(in results), cancellationToken);
```
def do not like this.  why can we not just pass this by ref?

it's also unclear to me what semantics this has for the dispose in this method.

sharwell replied:
> why can we not just pass this by ref?

Cannot take a `ref` to a `using` variable.

> it's also unclear to me what semantics this has for the dispose in this method.

It's extremely clear because both `ref` and `in` are specified. This is a reference to the same storage location as the using variable.

CyrusNajmabadi replied:
> Cannot take a ref to a using variable.


bleaghhhhhhhhhhhh....

The 'Unsafe' there really stands out like a sore thumb.  Is there any way we can have some sort of method on this like `ref result.PassAsRef()` to be a simpler and clearer pattern for everyone to follow?  It would still do this, just under the covers so it doesn't seem like we're explicitly doing unsafe awfulness at every initial pass site?

CyrusNajmabadi replied:
Aside: would it make sense to have an analyzer that forced all passing of this type ot be either by 'ref' or by 'in'?  i'm a bit worried about code forgetting to do that, and then randomly working or not depending on if it's the arraybuilder that values are being added to, or the inline-items.   Forcing ref/in would prevent htat.

sharwell replied:
> would it make sense to have an analyzer that forced all passing of this type ot be either by 'ref' or by 'in'?

This type is not copyable with analyzer enforcement. Attempting to pass by value will fail the build.

sharwell replied:
> Is there any way we can have some sort of method on this like `ref result.PassAsRef()` to be a simpler and clearer pattern for everyone to follow?

Unfortunately it's not possible to distinguish between the cases we want (referencing a `using` variable) and cases where we really don't want to write to the value (e.g. referencing an `in` parameter).

CyrusNajmabadi replied:
awesome.  i didn't know we had that, but that makes me feel *much* better here as i really care most about people bein able to properly use these types wih confidence.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs:
```diff
@@ -458,7 +444,7 @@ private static int BinarySearch(string concatenatedNames, ImmutableArray<Node> n
         // returns all the symbols in the container corresponding to the node
         private void Bind(
-            int index, INamespaceOrTypeSymbol rootContainer, ArrayBuilder<ISymbol> results, CancellationToken cancellationToken)
+            int index, INamespaceOrTypeSymbol rootContainer, ref TemporaryArray<ISymbol> results, CancellationToken cancellationToken)
```
needing to pass by ref seems worrying (unless we have an analyzer to catch this) as i can easily see someone accidentally passing by val and getting broken behavior.

sharwell replied:
The type is marked `NonCopyable`. The analyzer will not let a copy be made intentionally or accidentally.

CyrusNajmabadi replied:
that's awesome!  can you link me to that analyzer?

sharwell replied:
The bulk of the implementation is in https://github.com/dotnet/roslyn-analyzers/blob/master/src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs. Some language-syntax-specific operations are in the derived types.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs:
```diff
@@ -138,14 +137,15 @@ async Task<T> CreateWithLoggingAsync()
-                writer.WriteInt32(node.NameSpan.Length);
-                writer.WriteInt32(node.ParentIndex);
+                writer.WriteInt32(group.Length);
+                writer.WriteString(group.Span[0].Name);
```
nit, consider swapping these.  in practically all other places we write "length" then sequence of items.  so here we would write: name, length, items.   That feels morally better to me than: length, name, items.

sharwell replied:
➡️ Reordered

CyrusNajmabadi replied:
thanks.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs:
```diff
@@ -181,26 +181,49 @@ public void WriteTo(ObjectWriter writer)
                 }
             }
+
+            // sortedNodes is an array of Node instances which is often sorted by Node.Name by the caller. This method
```
often?

sharwell replied:
It's certainly not a precondition, and the algorithm will work correctly whether or not it's sorted.

CyrusNajmabadi replied:
when does it not happen?

this is one of those cases where i would prefer our code be stricter and precise about how it expects things ot be working when dealing entirely with internal data.  If we always ensure this is sorted, we can state as such and throw if not.  if we don't always ensure this, we can doc when this isn't the case.  Such documentation serves to better explain our own beliefs aobut the system for the future, as well as more easily catch when our expectations are potentially broken in the future.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs:
```diff
@@ -181,26 +181,49 @@ public void WriteTo(ObjectWriter writer)
+
+                var startIndex = 0;
+                var currentName = sortedNodes[0].Name;
+                for (var i = 0; i < sortedNodes.Length; i++)
```
consider starting at 1.  i found it slightly harder to convince myself of correctness in the case where we started on hte same item we'd already looked at and cached information for.

sharwell replied:
➡️ Done

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs:
```diff
@@ -181,26 +181,49 @@ public void WriteTo(ObjectWriter writer)
+                    var node = sortedNodes[i];
+                    if (node.Name != currentName)
+                    {
+                        yield return sortedNodes.AsMemory()[startIndex..i];
```
nit: should we hoist out the AsMemory view higher up?  i think it would help make this, and the below call nicer.

sharwell replied:
Now moved up

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs:
```diff
@@ -181,26 +181,49 @@ public void WriteTo(ObjectWriter writer)
-                    nodes.Add(new Node(new TextSpan(start, length), parentIndex));
+                    var groupCount = reader.ReadInt32();
+                    var name = reader.ReadString();
+                    for (var i = 0; i < groupCount; i++)
```
seeing this def makes me prefer the idea of having this be `name, length, items`

sharwell replied:
➡️ Tracking this under https://github.com/dotnet/roslyn/pull/48896#discussion_r513009479

CyrusNajmabadi commented on src/Workspaces/CoreTest/Collections/TemporaryArrayTests.cs:
```diff
@@ -0,0 +1,171 @@
+        {
+            Assert.Equal(0, TemporaryArray<int>.Empty.Count);
+            TemporaryArray<int>.Empty.Add(0);
+            Assert.Equal(0, TemporaryArray<int>.Empty.Count);
```
👍

CyrusNajmabadi commented on src/Workspaces/CoreTest/Collections/TemporaryArrayTests.cs:
```diff
@@ -0,0 +1,171 @@
+
+        [Theory]
+        [CombinatorialData]
+        public void TestAddRange([CombinatorialRange(0, 6)] int initialItems, [CombinatorialRange(0, 6)] int addedItems)
```
very cool :)

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,308 @@
+        /// This field is initialized to non-<see langword="null"/> the first time the <see cref="TemporaryArray{T}"/>
+        /// needs to store more than four elements. From that point, <see cref="_builder"/> is used instead of inline
+        /// elements, even if items are removed to make the result smaller than four elements.
+        /// </remarks>
```
👍

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,308 @@
+                if ((uint)index >= _count)
+                    ThrowIndexOutOfRangeException();
+
+                return Unsafe.Add(ref Unsafe.AsRef(in _item0), index);
```
are all these ops safe for any sort of T?   your tests all tested unmanaged types.  I admit i'm a bit wary about this for normal managed ref types.

sharwell replied:
I can add an assertion to a static constructor that `Unsafe.Add(in _item0, 1)` gives the address of `_item1`.

@tannergooding is the one who originally pointed out this form.

tannergooding replied:
Its safe for blittable data. I don't believe the VM will reorder for reference types either when all fields are T, but someone on the runtime team can likely confirm.

@jkotas?

jkotas replied:
If you go by the spec, the problem with this is that the may choose to enregister everything (AggressiveInlining on the method helps with that), it won't see any references of `_item1`, `_item2`, ... , and drops the storage for them.

The JIT has compatibility quirks to make these sort of cases work because of you are not the first one using this hack. I do not know the exact conditions when these compatibility quirks kick in.

You can make this more reliable by marking these methods non-inlineable.

sharwell replied:
@jkotas I have GC.KeepAlive references to all four fields in the Dispose method, which should directly address this

sharwell replied:
I'll switch to a switch expression for now since we need the rest of this win, and we can come back to this separately

jkotas replied:
> I have GC.KeepAlive references to all four fields in the Dispose method, which should directly address this

It is not enough to guarantee that the values are going to be spilled onto stack sequentially.

CyrusNajmabadi replied:
@sharwell  can we not just switch on the _count and get/set the appropriate item?  i'm struggling to think of a case where that would actuallybe a problem in the ide scenarios.

sharwell replied:
> can we not just switch on the _count and get/set the appropriate item?

Already done in accordance with https://github.com/dotnet/roslyn/pull/48896#discussion_r513073184

> i'm struggling to think of a case where that would actuallybe a problem in the ide scenarios.

The difference was _obvious_ in the benchmarks for https://github.com/tunnelvisionlabs/dotnet-trees/pull/102.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,308 @@
+            GC.KeepAlive(_item2);
+            GC.KeepAlive(_item3);
+
+            Interlocked.Exchange(ref _builder, null)?.Free();
```
can you either null out these four items, or explicitly state that we don't think it's necessary for yadda yadda.

sharwell replied:
➡️ Commented

CyrusNajmabadi replied:
thanks.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,308 @@
+        {
+            if (_builder is not null)
+            {
+                // Keep using a builder even if we now fit in inline storage to avoid churn in the object pools.
```
👍

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,308 @@
+            for (var i = 0; i < _count; i++)
+            {
+                builder.Add(this[i]);
+                this[i] = default!;
```
👍  up to you if we should explicitly call this out

sharwell replied:
I wrapped it in `RuntimeHelpers.IsReferenceOrContainsReferences<T>()` so it's clearer why this is getting set.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,308 @@
+        private static void ThrowIndexOutOfRangeException()
+            => throw new IndexOutOfRangeException();
+
+        public struct Enumerator
```
did i miss tests of this type in your test suite?

sharwell replied:
Added a test for this

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,323 @@
+        private T _item0;
+
+#pragma warning disable IDE0044 // Add readonly modifier
+#pragma warning disable CS0649 // Field 'name' is never assigned to, and will always have its default value
```
do we need this anymore, now that you directly access these by name?

sharwell replied:
➡️ No. Will remove and leave it for the follow-up.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,323 @@
+                    1 => _item1 = value,
+                    2 => _item2 = value,
+                    _ => _item3 = value,
+                };
```
i hate it :D

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,323 @@
+
+#if NETCOREAPP
+                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+#endif
```
cute :)

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,323 @@
+        private static void ThrowIndexOutOfRangeException()
+            => throw new IndexOutOfRangeException();
+
+        public struct Enumerator
```
aside, would it make sense for this itself to be a ref-struct?   in the future that might also be nice if it then held onto 'array' as a ref-field.  I honestly don't know.   Curious your thoughts here as you're more attuned to this sort of space.

sharwell replied:
> aside, would it make sense for this itself to be a ref-struct?

No. Non-copyable and stack storage are independent concepts. `ref struct` cannot be used in an `async` method, which would severely impair use of this type.

CyrusNajmabadi reviewed (approved):
very nice!

sharwell commented:
@jaredpar This just needs a final sign-off from you now

alrz commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs:
```diff
@@ -0,0 +1,313 @@
+    /// <typeparam name="T">The type of elements stored in the collection.</typeparam>
+    [NonCopyable]
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct TemporaryArray<T> : IDisposable
```
@sharwell I'm wondering if it is at all possible / preferable to use InlineArray for this?

sharwell replied:
@alrz `InlineArray` wouldn't be a complete replacement for this, but it would certainly be a viable replacement for the `_item0`...`_item3` fields.
